src/Tools/dotnet-dsrouter/DiagnosticsServerRouterCommands.cs

```diff
@@ -45,254 +45,274 @@ public void OnRouterStopped()
+            {
+                return new WebSocketServer.WebSocketServerImpl(runner.LogLevel);
+            });
 
-            using var factory = new LoggerFactory();
-            factory.AddConsole(logLevel, false);
 
-            Launcher.SuspendProcess = true;
-            Launcher.ConnectMode = false;
-            Launcher.Verbose = logLevel != LogLevel.Information;
-            Launcher.CommandToken = token;
+            return await runner.CommonRunLoop((logger, launcherCallbacks, linkedCancelToken) =>
+            {
+                NetServerRouterFactory.CreateInstanceDelegate webSocketServerRouterFactory = WebSocketServerRouterFactory.CreateDefaultInstance;
 
-            var logger = factory.CreateLogger("dotnet-dsrouter");
+                if (string.IsNullOrEmpty(ipcServer))
+                    ipcServer = GetDefaultIpcServerPath(logger);
 
-            TcpClientRouterFactory.CreateInstanceDelegate tcpClientRouterFactory = TcpClientRouterFactory.CreateDefaultInstance;
-            if (!string.IsNullOrEmpty(forwardPort))
+                var routerTask = DiagnosticsServerRouterRunner.runIpcServerTcpServerRouter(linkedCancelToken.Token, ipcServer, webSocket, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, webSocketServerRouterFactory, logger, launcherCallbacks);
+                return routerTask;
+            }, token);
+        }
+
+        class IpcClientWebSocketServerRunner : SpecificRunnerBase
+        {
+            public IpcClientWebSocketServerRunner(string verbose) : base(verbose) { }
+
+            public override void ConfigureLauncher(CancellationToken cancellationToken)
             {
-                if (string.Compare(forwardPort, "android", StringComparison.OrdinalIgnoreCase) == 0)
-                {
-                    tcpClientRouterFactory = ADBTcpClientRouterFactory.CreateADBInstance;
-                }
-                else if (string.Compare(forwardPort, "ios", StringComparison.OrdinalIgnoreCase) == 0)
-                {
-                    tcpClientRouterFactory = USBMuxTcpClientRouterFactory.CreateUSBMuxInstance;
-                }
-                else
-                {
-                    logger.LogError($"Unknown port forwarding argument, {forwardPort}. Ignoring --forward-port argument.");
-                }
+                Launcher.SuspendProcess = true;
+                Launcher.ConnectMode = true;
+                Launcher.Verbose = LogLevel != LogLevel.Information;
+                Launcher.CommandToken = cancellationToken;
             }
+        }
 
-            var routerTask = DiagnosticsServerRouterRunner.runIpcClientTcpClientRouter(linkedCancelToken.Token, ipcClient, tcpClient, runtimeTimeout == Timeout.Infinite ? runtimeTimeout : runtimeTimeout * 1000, tcpClientRouterFactory, logger, Launcher);
+        public async Task<int> RunIpcClientWebSocketServerRouter(CancellationToken token, string ipcClient, string webSocket, int runtimeTimeout, string verbose)
```

Comment:
I guess this should use  checkLoopbackOnly as well, we need to warn if you try to bind anything except loopback interface.